actions/watch.ts

```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
-          );
+  // Update the registry
+  await registry.write();
+};
+
+async function _checkForAndPlaceOrder(
+  owner: string,
+  network: string,
+  conditionalOrder: ConditionalOrder,
+  contract: ComposableCoW,
+  chainContext: ChainContext
+): Promise<{ deleteConditionalOrder: boolean }> {
+  try {
+    const { order, signature } = await getTradeableOrderWithSignature(
+      owner,
+      conditionalOrder,
+      contract
+    );
+
+    const orderToSubmit: Order = {
+      ...order,
+      kind: kindToString(order.kind),
+      sellTokenBalance: balanceToString(order.sellTokenBalance),
+      buyTokenBalance: balanceToString(order.buyTokenBalance),
+    };
+
+    // calculate the orderUid
+    const orderUid = getOrderUid(network, orderToSubmit, owner);
+
+    // if the orderUid has not been submitted, or filled, then place the order
+    if (!conditionalOrder.orders.has(orderUid)) {
+      await placeOrder(
+        { ...orderToSubmit, from: owner, signature },
+        chainContext.api_url
+      );
 
-          conditionalOrder.orders.set(orderUid, OrderStatus.SUBMITTED);
-        } else {
+      conditionalOrder.orders.set(orderUid, OrderStatus.SUBMITTED);
+    } else {
+      const orderStatus = conditionalOrder.orders.get(orderUid);
+      console.log(
+        `OrderUid ${orderUid} status: ${
+          orderStatus ? formatStatus(orderStatus) : "Not found"
+        }`
+      );
+    }
+  } catch (e: any) {
+    if (e.code === Logger.errors.CALL_EXCEPTION) {
+      switch (e.errorName) {
+        case "OrderNotValid":
+          // The conditional order has not expired, or been cancelled, but the order is not valid
+          // For example, with TWAPs, this may be after `span` seconds have passed in the epoch.
+          return { deleteConditionalOrder: false };
+        case "SingleOrderNotAuthed":
           console.log(
-            `OrderUid ${orderUid} status: ${conditionalOrder.orders.get(
-              orderUid
-            )}`
+            `Single order on safe ${owner} not authed. Unfilled orders:`
           );
-        }
-      } catch (e: any) {
-        if (e.code === Logger.errors.CALL_EXCEPTION) {
-          switch (e.errorName) {
-            case "OrderNotValid":
-              // The conditional order has not expired, or been cancelled, but the order is not valid
-              // For example, with TWAPs, this may be after `span` seconds have passed in the epoch.
-              continue;
-            case "SingleOrderNotAuthed":
-              console.log(
-                `Single order on safe ${owner} not authed. Unfilled orders:`
-              );
-            case "ProofNotAuthed":
-              console.log(
-                `Proof on safe ${owner} not authed. Unfilled orders:`
-              );
-          }
-          printUnfilledOrders(conditionalOrder.orders);
-          console.log("Removing conditional order from registry");
-          conditionalOrders.delete(conditionalOrder);
```

Comment:
I don't do the deletion here anymore, i just signal to the main loop this conditional order should be removed, so:
-  we don't mutate the `conditionalOrders` (we are iterating over them)
- Its super easy to miss this logic when reading. I just made it explicit so is part of this method signature to tell the main loop when it should delete an order

Reply:
Yep, this makes it significantly more readable / understandable. Good work.